File: trunk/src/ruby/damagecontrol/FakeClock.rb
require 'damagecontrol/Clock'

module DamageControl
	
	class FakeClock < Clock
		def initialize
			@time = 0
			@has_waiters = Latch.new
			@time_changed = Latch.new
		end
		
		def change_time (newtime)
			@time = newtime
			@time_changed.release()
			@time_changed = Latch.new
		end
		
		def wait_for_waiters (timeout=-1)
			@has_waiters.wait()
		end
		
		def current_time
			@time
		end
		
		def wait_until (wait_until_time)
			@has_waiters.release()
			while current_time < wait_until_time
				@time_changed.wait()
			end
		end
	end
	
end

File: trunk/src/test/damagecontrol/FakeClock.java
package damagecontrol;

import junit.framework.Assert;
import EDU.oswego.cs.dl.util.concurrent.Latch;

public class FakeClock implements Clock {
    private long currentTimeMillis;
    private Latch hasWaiterLatch = new Latch();
    private Latch currentTimeChangedLatch = new Latch();

    public void waitUntil(long timeToWaitUntil) {
        hasWaiterLatch.release();

        while (!Thread.currentThread().isInterrupted()) {
            synchronized (this) {
                if (currentTimeMillis >= timeToWaitUntil) {
                    return;
                }
            }
            try {
                currentTimeChangedLatch.attempt(1000);
            } catch (InterruptedException e) {
            }
        }
    }

    public long currentTimeMillis() {
        return currentTimeMillis;
    }

    public synchronized void changeTime(int currentTimeMillis) {
        this.currentTimeMillis = currentTimeMillis;
        currentTimeChangedLatch.release();
        currentTimeChangedLatch = new Latch();
    }

    public synchronized void waitForWaiter() {
        try {
            Assert.assertTrue("timeout, no waiter arrived", hasWaiterLatch.attempt(1000));
        } catch (InterruptedException e) {
            Assert.fail();
        }
    }
}
